src/test/lib/until_test.ts

```diff
@@ -22,20 +22,56 @@ const assert = chai.assert;
 
 suite('until', () => {
 
+  let container: HTMLDivElement;
+  let deferred: Deferred<string>;
+
+  setup(() => {
+    container = document.createElement('div');
+    deferred = new Deferred<string>();
+  });
+
   test('displays defaultContent immediately', async () => {
-    const container = document.createElement('div');
-    let resolve: (v: any) => void;
-    const promise = new Promise((res, _) => {
-      resolve = res;
-    });
     render(
-        html`<div>${until(promise, html`<span>loading...</span>`)}</div>`,
+        html
+        `<div>${until(deferred.promise, html`<span>loading...</span>`)}</div>`,
         container);
     assert.equal(container.innerHTML, '<div><span>loading...</span></div>');
-    resolve!('foo');
-    await promise;
-    await new Promise((r) => setTimeout(() => r()));
+
+    deferred.resolve('foo');
+    await deferred.promise;
     assert.equal(container.innerHTML, '<div>foo</div>');
   });
 
+  test('renders new Promise over existing Promise', async () => {
+    const t = (v: any) =>
+        html`<div>${until(v, html`<span>loading...</span>`)}</div>`;
+    render(t(deferred.promise), container);
+    assert.equal(container.innerHTML, '<div><span>loading...</span></div>');
+
+    const deferred2 = new Deferred<string>();
+    render(t(deferred2.promise), container);
+    assert.equal(container.innerHTML, '<div><span>loading...</span></div>');
+
+    deferred2.resolve('bar');
+    await deferred2.promise;
+    assert.equal(container.innerHTML, '<div>bar</div>');
+
+    deferred.resolve('foo');
+    await deferred.promise;
+    assert.equal(container.innerHTML, '<div>bar</div>');
+  });
+
 });
+
+class Deferred<T> {
```

Comment:
Consider splitting this out into a test helper library.

Reply:
done